concordium-std/src/state_btree.rs

```diff
@@ -0,0 +1,1481 @@
+/// }
+/// ```
+/// Or alternatively
+/// ```no_run
+/// # use concordium_std::*;
+/// # struct MyState<S: HasStateApi = StateApi> {
+/// #    inner: StateBTreeMap<u64, u64, S>
+/// # }
+/// fn correct_replace(state_builder: &mut StateBuilder, state: &mut MyState) {
+///     let old_map = mem::replace(&mut state.inner, state_builder.new_btree_map());
+///     old_map.delete()
+/// }
+/// ```
+pub struct StateBTreeMap<K, V, S, const M: usize = 8> {
+    pub(crate) map:         StateMap<K, V, S>,
+    pub(crate) ordered_set: StateBTreeSet<K, S, M>,
+}
+
+impl<const M: usize, K, V, S> StateBTreeMap<K, V, S, M> {
+    /// Insert a key-value pair into the map.
+    /// Returns the previous value if the key was already in the map.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V>
+    where
+        S: HasStateApi,
+        K: Serialize + Ord,
+        V: Serial + DeserialWithState<S>, {
+        let old_value_option = self.map.insert_borrowed(&key, value);
+        if old_value_option.is_none() && !self.ordered_set.insert(key) {
+            // Inconsistency between the map and ordered_set.
+            crate::trap();
+        }
+        old_value_option
+    }
+
+    /// Remove a key from the map, returning the value at the key if the key was
+    /// previously in the map.
+    ///
+    /// *Caution*: If `V` is a [StateBox], [StateMap], then it is
+    /// important to call [`Deletable::delete`] on the value returned when
+    /// you're finished with it. Otherwise, it will remain in the contract
+    /// state.
+    #[must_use]
+    pub fn remove_and_get(&mut self, key: &K) -> Option<V>
+    where
+        S: HasStateApi,
+        K: Serialize + Ord,
+        V: Serial + DeserialWithState<S> + Deletable, {
+        let v = self.map.remove_and_get(key);
+        if v.is_some() && !self.ordered_set.remove(key) {
+            // Inconsistency between the map and ordered_set.
+            crate::trap();
+        }
+        v
+    }
+
+    /// Remove a key from the map.
+    /// This also deletes the value in the state.
+    pub fn remove(&mut self, key: &K)
+    where
+        S: HasStateApi,
+        K: Serialize + Ord,
+        V: Serial + DeserialWithState<S> + Deletable, {
+        if self.ordered_set.remove(key) {
+            self.map.remove(key);
+        }
+    }
+
+    /// Get a reference to the value corresponding to the key.
+    pub fn get(&self, key: &K) -> Option<StateRef<V>>
+    where
+        K: Serialize,
+        S: HasStateApi,
+        V: Serial + DeserialWithState<S>, {
+        if self.ordered_set.is_empty() {
+            None
+        } else {
+            self.map.get(key)
+        }
+    }
+
+    /// Get a mutable reference to the value corresponding to the key.
+    pub fn get_mut(&mut self, key: &K) -> Option<StateRefMut<V, S>>
+    where
+        K: Serialize,
+        S: HasStateApi,
+        V: Serial + DeserialWithState<S>, {
+        if self.ordered_set.is_empty() {
+            None
```

Comment:
We don't need it, but it is a small optimization.
Since the `ordered_set` is tracking the length, we can complete avoid the host call for looking up the key if the collection is empty.